doc/docbook/nmsgtool.1
'\" t
.\"     Title: nmsgtool
.\"    Author: [FIXME: author] [see http://www.docbook.org/tdg5/en/html/author]
.\" Generator: DocBook XSL Stylesheets v1.79.2 <http://docbook.sf.net/>
.\"      Date: 10/02/2024
.\"    Manual:  
.\"    Source:    
.\"  Language: English
.\"
.TH "NMSGTOOL" "1" "10/02/2024" "" ""
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
nmsgtool \- command line interface to libnmsg
.SH "SYNOPSIS"
.HP \w'\fBnmsgtool\fR\ 'u
\fBnmsgtool\fR [\fIOPTIONS\fR...]
.SH "DESCRIPTION"
.PP
\fBnmsgtool\fR
is the command line interface to libnmsg, the reference implementation of the NMSG binary structured message interchange format\&.
.PP
The NMSG format is an efficient encoding of typed, structured data into payloads which are packed into containers which can be transmitted over the network or stored to disk\&. Each payload is associated with a specific message schema\&. Modules implementing a certain message schema along with functionality to convert between binary and presentation formats can be loaded at runtime by libnmsg\&.
\fBnmsgtool\fR
provides a command line interface to control the transmission, storage, creation, and conversion of NMSG payloads\&.
.PP
\fBnmsgtool\fR
is a thin wrapper around libnmsg\*(Aqs I/O engine\&. In order to run,
\fBnmsgtool\fR
needs to open at least one input and at least one output\&. The supported input types are:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Reading binary NMSG data from a file or socket\&. See the
\fB\-r\fR,
\fB\-l\fR
options\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Reassembled IP datagrams from a network interface or
\fBpcap-savefile\fR(5)
file\&. See the
\fB\-i\fR,
\fB\-p\fR
options\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Reading ASCII presentation form data from a file\&. See the
\fB\-f\fR
option\&.
.RE
.PP
Once input data has been read or converted to NMSG format internally with the help of an external module (in the case of pcap or ASCII data), it is striped or mirror across one or more outputs\&. The supported output types are:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Writing binary NMSG data to a file or socket\&. See the
\fB\-w\fR,
\fB\-s\fR
options\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Writing ASCII presentation form data to a file\&. See the
\fB\-o\fR
option\&.
.RE
.PP
Reading or writing data in a non\-NMSG format requires the use of an external module (called an "nmsgpb module") to convert to or from NMSG format\&.
\fBnmsgtool\fR
selects an nmsgpb module based on a vendor ID and message type\&. For input data, these fields must be set with the
\fB\-V\fR
and
\fB\-T\fR
options\&.
.PP
If one or more inputs have been specified but no outputs have been specified, an implicit output of presentation form data to stdout will be assumed\&.
.PP
See the nmsg developer documentation for details on the NMSG wire format, how to interface with the nmsg library, and how to extend nmsg with new message types\&.
.SH "OPTIONS"
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Display help text and exit\&.
.RE
.PP
\fB\-d\fR, \fB\-\-debug\fR
.RS 4
Increment debugging level\&.
\fB\-dd\fR
is verbose and
\fB\-dddd\fR
is very verbose\&.
If the
\fBNMSG_KAFKA_LOG_LEVEL\fR
environment variable is set and a Kafka input/output is used, Kafka will log non-errors at the specified logging level\&.
.RE
.PP
\fB\-v\fR \fIversion\fR
.RS 4
Print the version number of \fBnmsgtool\fR.
.RE
.PP
\fB\-V\fR \fIvendor\fR, \fB\-\-vendor\fR \fIvendor\fR
.RS 4
Set the vendor field of generated NMSG payloads to the vendor identified by
\fIvname\fR\&. This is a symbolic string whose allowed values are determined at runtime when nmsgpb modules are loaded\&.
.RE
.PP
\fB\-T\fR \fImsgtype\fR, \fB\-\-msgtype\fR \fImsgtype\fR
.RS 4
Set the message type field of generated NMSG payloads to the message type identified by
\fImsgtype\fR\&. This is a symbol string whose allowed values are determined at runtime when nmsgpb modules are loaded\&.
.RE
.PP
\fB\-B\fR \fIbyterate\fR, \fB\-\-byterate\fR \fIbyterate\fR
.RS 4
Set the maximum bytes-per-second that libnmsg will process from file inputs.
.RE
.PP
\fB\-D\fR \fIdaemon\fR
.RS 4
Fork \fBnmsgtool\fR into the background as a daemon process.
.RE
.PP
\fB\-P\fR \fIpidfile\fR, \fB\-\-pidfile\fR \fIpidfile\fR
.RS 4
Write \fBnmsgtool\fR's process ID into a file identified by \fIpidfile\fR.  When 
\fBnmsgtool\fR exits, this file's contents will be erased.
.RE
.PP
\fB\-U\fR \fIusername\fR, \fB\-\-username\fR \fIusername\fR
.RS 4
Attempt to drop root privileges and run \fBnmsgtool\fR as user \fIusername\fR.  
If the current user is \fIusername\fR, \fBnmsgtool\fR will exit.
.RE
.PP
\fB\-e\fR \fIendline\fR, \fB\-\-endline\fR \fIendline\fR
.RS 4
Set the string to be used for the end\-of\-line continuation separator\&. By default this value is set to the newline character\&. This option understands the escape sequences
\fB\*(Aq\en\*(Aq\fR
(newline),
\fB\*(Aq\et\*(Aq\fR
(tab), and
\fB\*(Aq\e\e\*(Aq\fR
(backslash)\&. Setting this option to
\fB\*(Aq \e\e\en\et\*(Aq\fR
will emulate the line continuation behavior of
\fBncaptool\fR\&.
.RE
.PP
\fB\-m\fR \fImtu\fR, \fB\-\-mtu\fR \fImtu\fR
.RS 4
Set the "maximum transmission unit" for writes to datagram socket outputs\&. Note that this option is somewhat misnamed since it does not refer to protocol data units of IP but to NMSG containers, so the additional small but bounded overhead of the UDP and NMSG headers need to be taken into account\&. Since NMSG\*(Aqs UDP transport is tuned for operation on jumbo Ethernet, this value defaults to
\fI8192\fR\&. To avoid IP fragmentation over standard Ethernet or Internet transport, this value should be lowered to around
\fI1280\fR
to
\fI1400\fR\&.
.sp
This option does not limit the maximum size of an NMSG payload because NMSG containers can be segmented into multiple datagrams\&.
.RE
.PP
\fB\-c\fR \fIcount\fR, \fB\-\-count\fR \fIcount\fR
.RS 4
Stop the process or (with
\fB\-k\fR
specified) reopen outputs after
\fIcount\fR
payloads have been processed\&. Note that each output keeps its own payload output counter and this value refers to that per\-output value and not to the sum of all output counters\&.
.RE
.PP
\fB\-t\fR \fIsecs\fR, \fB\-\-interval\fR \fIsecs\fR
.RS 4
Stop the process or (with
\fB\-k\fR
specified) reopen outputs on a time interval modulo the
\fIsecs\fR
parameter\&. For example,
\fI\-t 3600\fR
means "on the hour"\&.
.RE
.PP
\fB\-R\fR, \fB\-\-randomize\fR
.RS 4
Randomize the initial offset within the interval that the process is stopped or outputs are reopened\&.
.RE
.PP
\fB\-k\fR \fIcmd\fR, \fB\-\-kicker\fR \fIcmd\fR
.RS 4
Make
\fB\-c\fR
and
\fB\-t\fR
continuous\&. In this mode output file names are suffixed with a timestamp and
\fBnmsgtool\fR
runs continuously, rotating output files as payload counts or time intervals expire\&.
\fIcmd\fR
specifies the command to run on output files after rotation\&. If
\fIcmd\fR
is set to the empty string \*(Aq\*(Aq, then no command is executed and only file rotation is performed\&.
.RE
.PP
\fB\-\-kafkakey\fR \fIfieldname\fR
.RS 4
Use the value of the named NMSG message field in its canonical representation as the key for payloads in the JSON format that are sent to Kafka\&. Can also be set via
\fBNMSG_KAFKA_KEY\fR
environment variable\&.
.RE
.PP
\fB\-\-readkafka\fR \fIkafka\fR
.RS 4
Read NMSG payloads in either binary or JSON format from a Kafka endpoint\&. The address
\fIkafka\fR
has format
\fBproto:topic[#partition|%group_id]@broker[:port][,offset]\fR\&. Either a partition number or a consumer group ID may be optionally supplied\&. Also optional is an offset consisting of either a numerical value or the string \*(Aqoldest\*(Aq or \*(Aqnewest\*(Aq in order to start retrieval at the oldest/newest messages in the Kafka topic\&. An example of a possible
\fIkafka\fR
endpoint is "nmsg:ch202#0@kafka\&.example\&.com:9092,3000" to indicate that nmsgtool shall read nmsg containers from topic "ch202" on partition 0 at offset 3000 from the Kafka broker at kafka\&.example\&.com, port 9092\&. Configuration for Kafka can be supplied through the
\fBNMSG_KAFKA_CONFIG\fR
environment variable\&. This variable accepts either key=value pairs separated by colons, or a path to a configuration file in INI format\&.
.RE
.PP
\fB\-b\fR \fIfilter\fR, \fB\-\-bpf\fR \fIfilter\fR
.RS 4
Filter pcap inputs (\fB\-p\fR
or
\fB\-i\fR) with the BPF expression
\fIfilter\fR\&. Note that libnmsg\*(Aqs pcap input is designed for IP datagrams and not network frames, so the filter specified by
\fB\-b\fR
will be munged internally into several forms in order to receive IP fragments and filter the reassembled IP datagrams\&. Internally, IPv4 and IPv6 packets are selected for processing, including those received in VLAN tagged frames\&.
.RE
.PP
\fB\-F\fR \fIdso\fR[,\fIparam\fR], \fB\-\-filter\fR \fIdso\fR[,\fIparam\fR]
.RS 4

ChangeLog
nmsg (1.3.3-1)

  * Adjust debug log levels for Kafka.

  * Add NMSG_KAFKA_FULL_LOGGING for increased verbosity.

nmsg (1.3.2-1)

  * Add configure --with-coverage feature and make targets "clean-coverage"
    and "report-coverage" for code coverage analysis.

  * Add more documentation to nmsgtool manual.

  * Remove unused dead code.

  * Lots of tests added and improved.

  * Add sample statsmod plugin (nmsg_stats1_sample) with parameters
    "out" to define output filename and "secs" to set the update
    frequency in seconds.

  * Various packet capture parsing fixes for DNSQR message type.

  * Fix bug where idle JSON and presentation inputs block SIGINT.

nmsg (1.3.1-1)

  * Rely on built-in librdkafka error handling.

nmsg (1.2.0)

  * Add support for Kafka i/o in JSON and binary forms.

  * Add support for prometheus metrics export.

  * Replace mutex protected counters and flags with stdatomic operations.

  * New command line options: --kafkakey, --readkafka, --writekafka, --promport

  * New functions: nmsg_input_open_kafka_endpoint(), nmsg_input_open_kafka_json(),
                   nmsg_output_open_kafka_endpoint(), nmsg_output_open_kafka_json()

nmsg (1.1.2)

  * Now allow nmsgtool --setgroup, --setoperator, and --setsource to
    work with --writepres and --writejson outputs.

  * Permit nmsgtool --setgroup and --setoperator to use a numerical
    value.

  * Unmatched group and operator aliases in presentation format
    now display as the numerical values instead of (null).

  * Ensure mis-running nmsgtool results in failure exit code.

  * Fix output buffering for JSON outputs.

  * Fix magic length calculation in nmsg_json.

  * Terminate strbuf data at initialization.


nmsg (1.1.1)

  * Update estimated container size to include space for sequencing info.

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Mon, 11 Sep 2023 20:21:09 +0000

nmsg (1.1.0)

  * Multiple performance improvements to JSON formatting.

  * Multiple performance and concurrency improvements in nmsg_io loop.

  * Add new dnsobs message type for raw passive DNS observations.

  * Format DNS messages as JSON in base:dnsqr, base:dnstap, and base:dnsobs.

  * Format base:encode payload in human readable form in JSON output.

  * Replace yajl with json-c for JSON parsing.

  * nmsgtool: exit with nonzero status when interrupted by signal.

  * Correct container size calculation when sequencing.

  * Address doxygen deprecations.

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Wed, 16 Aug 2023 18:43:40 +0000

nmsg (1.0.1)

  * nmsgtool: Fix use-after-free bug in statistics debugging
    (seen on Debian 10).

  * nmsg/base/dnsqr.c: Initialize len in dnsqr_hash to squash
    code-checker uninitialized warning.

  * Fix assert conditions with side effects for builds with -DNDEBUG.

nmsg (1.0.0)

  * Replace libxs with zeromq and associated calls.

  * API functions nmsg_input_open_xs(), nmsg_input_open_xs_endpoint(),
    nmsg_io_add_input_xs_channel(), nmsg_output_open_xs(), and
    nmsg_output_open_xs_endpoint() are deprecated and replaced with
    respectively: nmsg_input_open_zmq(), nmsg_input_open_zmq_endpoint(),
    nmsg_io_add_input_zmq_channel(), nmsg_output_open_zmq(), and
    nmsg_output_open_zmq_endpoint().

  * Change format of unknown rrtypes in nmsgtool JSON output to "TYPE" followed
    by number.

  * Update nmsgtool command-line options: --readzchan replaces --readzsock,
    --writezsock replaces --writexsock.

nmsg (0.15.1)

  * Fix output stats collection.

  * Plugins now found by default at $libdir/nmsg instead of lib/nmsg.
    This fixes problem where plugins were not found with recent Debian
    packages built with --libdir configure option (used to install
    libraries in architecture specific directory). And
    ./configure --with-plugindir renamed to --with-pluginsdir.

  * pkgconfig also depend on libprotobuf-c headers (for third-party
    uses of libnmsg with pkgconfig).

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Thu, 23 Jan 2020 13:20:13 +0000

nmsg (0.15.0)

  * Add compile-time and runtime version info to API: NMSG_LIBRARY_VERSION
    and nmsg_get_version(), respectively.

  * Fixed many compiler check warnings.

  * Added multiple unit tests.

  * Added runtime-tunable environment variables NMSG_GRALIAS_FILE and
    NMSG_OPALIAS_FILE.

  * Ensure nmsg_sock_parse_sockspec() returns an error if an invalid
    port range is supplied.

  * Fixed nmsg_io filtering default policy logic. (Previously, if the
    final filter returned "declined", that result was returned in place
    of the configured default policy.)

  * Make nmsg_output_set_buffered() able to unbuffer nmsg_output_type_json
    objects.

  * Fix bugs caught by static analysis (cppcheck).

  * For protobuf compatibility, renamed all protobuf-c generated enum

src/nmsgtool.c
/*
 * Copyright (c) 2023-2024 DomainTools LLC
 * Copyright (c) 2008-2021 by Farsight Security, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* Import. */

#include <assert.h>
#include <errno.h>
#include <inttypes.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <nmsg.h>

#include "nmsgtool.h"
#include "kickfile.h"

/* Globals. */

static nmsgtool_ctx ctx;

static argv_t args[] = {
	{ 'b',	"bpf",
		ARGV_CHAR_P,
		&ctx.bpfstr,
		"filter",
		"filter pcap inputs with this bpf" },

	{ 'B', "byterate",
		ARGV_INT,
		&ctx.byte_rate,
		"byterate",
		"ingress byte rate limit for file input" },

	{ 'c',	"count",
		ARGV_INT,
		&ctx.count,
		"count",
		"stop or reopen after count payloads output" },

	{ 'C', "readchan",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.r_channel,
		"channel",
		"read nmsg data from socket(s)" },

	{ 'd',	"debug",
		ARGV_INCR,
		&ctx.debug,
		NULL,
		"increment debugging level" },

	{ 'D', "daemon",
		ARGV_BOOL,
		&ctx.daemon,
		NULL,
		"fork into background" },

	{ 'e', "endline",
		ARGV_CHAR_P,
		&ctx.endline,
		"endline",
		"continuation separator" },

	{ 'F',	"filter",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.filters,
		"dso[,param]",
		"filter nmsg payloads with module" },

	{ '\0',	"getgroup",
		ARGV_CHAR_P,
		&ctx.get_group_str,
		"grname",
		"only process payloads with this group name" },

	{ '\0', "getoperator",
		ARGV_CHAR_P,
		&ctx.get_operator_str,
		"opname",
		"only process payloads with this operator name" },

	{ '\0', "getsource",
		ARGV_CHAR_P,
		&ctx.get_source_str,
		"sonum",
		"only process payloads with this source value" },

	{ 'h',	"help",
		ARGV_BOOL,
		&ctx.help,
		NULL,
		"display help text and exit" },

	{ 'i',	"readif",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.r_pcapif,
		"if[+][,snap]",
		"read pcap data from interface ('+' = promisc)" },


	{ 'j', "readjson",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.r_json,
		"file",
#ifdef HAVE_JSON_C
		"read json format data from file" },
#else /* HAVE_JSON_C */
		"read json format data from file (no support)" },
#endif /* HAVE_JSON_C */

	{ 'J', "writejson",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.w_json,
		"file",
		"write json format data to file" },

	{ 'k',	"kicker",
		ARGV_CHAR_P,
		&ctx.kicker,
		"cmd",
		"make -c, -t continuous; run cmd on new files" },
	{'\0', "kafkakey",
		ARGV_CHAR_P,
		&ctx.kafka_key_field,
		"fieldname",
#if defined(HAVE_LIBRDKAFKA) && defined(HAVE_JSON_C)
		"nmsg field for Kafka producer key" },
#else /* defined(HAVE_LIBRDKAFKA) && defined(HAVE_JSON_C) */
		"nmsg field for Kafka producer key (no support)" },
#endif /* defined(HAVE_LIBRDKAFKA) && defined(HAVE_JSON_C) */


	{'\0', "readkafka",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.r_kafka,
		"kafka",
#ifdef HAVE_LIBRDKAFKA
#ifdef HAVE_JSON_C
		"read nmsg data from Kafka (binary or json)" },
#else /* HAVE_JSON_C */
		"read nmsg containers from Kafka topic" },
#endif /* HAVE_JSON_C */
#else /* HAVE_LIBRDKAFKA */
		"read nmsg data from Kafka topic (no support)" },
#endif /* HAVE_LIBRDKAFKA */

	{ 'l', "readsock",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.r_sock,
		"so",
		"read nmsg data from socket (addr/port)" },

	{ 'L', "readzsock",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.r_zsock,
		"zep",
#ifdef HAVE_LIBZMQ
		"read nmsg data from ZMQ endpoint" },
#else /* HAVE_LIBZMQ */
		"read nmsg data from ZMQ endpoint (no support)" },
#endif /* HAVE_LIBZMQ */

	{ 'm', "mtu",
		ARGV_INT,
		&ctx.mtu,
		"mtu",
		"MTU for datagram socket outputs" },

	{ 'M', "statsmod",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.statsmods,
		"dso[,params]",
		"Export IO metrics with module" },

	{ '\0', "mirror",
		ARGV_BOOL,
		&ctx.mirror,
		NULL,
		"mirror payloads across data outputs" },

	{ 'o', "writepres",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.w_pres,
		"file",
		"write pres format data to file" },

	{ 'p',	"readpcap",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.r_pcapfile,
		"file",
		"read pcap data from file" },

	{ 'P', "pidfile",
		ARGV_CHAR_P,
		&ctx.pidfile,
		"file",
		"write PID into file" },

	{ '\0', "policy",
		ARGV_CHAR_P,
		&ctx.filter_policy,
		"ACCEPT|DROP",
		"default filter chain policy" },

	{ 'r', "readnmsg",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.r_nmsg,
		"file",
		"read nmsg data from file" },


	{ 'R', "randomize",
		ARGV_BOOL,
		&ctx.interval_randomized,
		NULL,
		"randomize beginning of -t interval" },

	{ 's', "writesock",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.w_sock,
		"so[,r[,f]]",
		"write nmsg data to socket (addr/port)" },

	{ 'S', "writezsock",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.w_zsock,
		"zep",
#ifdef HAVE_LIBZMQ
		"write nmsg data to ZMQ endpoint" },
#else /* HAVE_LIBZMQ */
		"write nmsg data to ZMQ endpoint (no support)" },
#endif /* HAVE_LIBZMQ */

	{ '\0',	"setgroup",
		ARGV_CHAR_P,
		&ctx.set_group_str,
		"grname",
		"set payload group to this value" },

	{ '\0',	"setoperator",
		ARGV_CHAR_P,
		&ctx.set_operator_str,
		"opname",
		"set payload operator to this value" },

	{ '\0',	"setsource",
		ARGV_CHAR_P,
		&ctx.set_source_str,
		"sonum",
		"set payload source to this value" },

	{ 't',	"interval",
		ARGV_INT,
		&ctx.interval,
		"secs",
		"stop or reopen after secs have elapsed" },

	{ 'T', "msgtype",
		ARGV_CHAR_P,
		&ctx.mname,
		"msgtype",
		"message type" },

	{ 'U', "username",
		ARGV_CHAR_P,
		&ctx.username,
		"user",
		"drop privileges and run as user" },

	{ '\0', "unbuffered",
		ARGV_BOOL,
		&ctx.unbuffered,
		NULL,
		"don't buffer writes to outputs" },

	{ 'v', "version",
		ARGV_BOOL,
		&ctx.version,
		NULL,
		"print version" },

	{ 'V', "vendor",
		ARGV_CHAR_P,
		&ctx.vname,
		"vendor",
		"vendor" },

	{ 'w', "writenmsg",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.w_nmsg,
		"file",
		"write nmsg data to file" },

	{ '\0', "writekafka",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.w_kafka,
		"kafka",
#ifdef HAVE_LIBRDKAFKA
#ifdef HAVE_JSON_C
		"write nmsg data to Kafka (binary or json)" },
#else /* HAVE_JSON_C */
		"write nmsg containers to to Kafka topic" },
#endif /* HAVE_JSON_C */
#else /* HAVE_LIBRDKAFKA */
		"write nmsg data to Kafka topic (no support)" },
#endif /* HAVE_LIBRDKAFKA */

	{ 'Z', "readzchan",
		ARGV_CHAR_P | ARGV_FLAG_ARRAY,
		&ctx.r_zchannel,
		"zchannel",
#ifdef HAVE_LIBZMQ
		"read nmsg data from ZMQ channels" },
#else /* HAVE_LIBZMQ */
		"read nmsg data from ZMQ channels (no support)" },
#endif /* HAVE_LIBZMQ */

	{ 'z', "zlibout",
		ARGV_BOOL,
		&ctx.zlibout,
		NULL,
		"compress nmsg output" },

	{ ARGV_LAST, 0, 0, 0, 0, 0 }
};


static void print_io_stats(nmsg_io_t);
static void io_close(struct nmsg_io_close_event *);
static void setup_signals(void);
static void signal_handler(int);

/* Functions. */

int main(int argc, char **argv) {
	nmsg_res res;

	/* parse command line arguments */
	argv_process(args, argc, argv);

	if (ctx.debug < 1)
		ctx.debug = 1;
	nmsg_set_debug(ctx.debug);
	res = nmsg_init();
	if (res != nmsg_res_success) {
		fprintf(stderr, "nmsgtool: unable to initialize libnmsg\n");
		return (EXIT_FAILURE);
	}
	if (ctx.debug >= 2)
#ifdef HAVE_LIBZMQ
		fprintf(stderr, "nmsgtool: version " VERSION "\n");
#else /* HAVE_LIBZMQ */
		fprintf(stderr, "nmsgtool: version " VERSION " (without libzmq support)\n");
#endif /* HAVE_LIBZMQ */

	ctx.statsmods_loaded = statsmod_vec_init(1);

	/* initialize the nmsg_io engine */
	ctx.io = nmsg_io_init();
	if (ctx.io == NULL) {
		fprintf(stderr, "Error: failed to initialize NMSG IO engine\n");
		exit(EXIT_FAILURE);
	}
	nmsg_io_set_close_fp(ctx.io, io_close);

	/* process arguments and load inputs/outputs into the nmsg_io engine */
	process_args(&ctx);

	setup_signals();

	/* run the nmsg_io engine */
	res = nmsg_io_loop(ctx.io);

	/* print stats, if requested */
	if (ctx.debug >= 2) {
		print_io_stats(ctx.io);
	}

	/* cleanup */
	if (ctx.pidfile != NULL) {
		if (unlink(ctx.pidfile) != 0) {
			fprintf(stderr, "nmsgtool: unlink() failed: %s\n",
				strerror(errno));
		}
	}

	for (size_t i = 0; i < statsmod_vec_size(ctx.statsmods_loaded); i++) {
		nmsg_statsmod_destroy(&statsmod_vec_data(ctx.statsmods_loaded)[i]);
	}
	statsmod_vec_destroy(&ctx.statsmods_loaded);

	nmsg_io_destroy(&ctx.io);
#ifdef HAVE_LIBZMQ
	if (ctx.zmq_ctx)
		zmq_term(ctx.zmq_ctx);
#endif /* HAVE_LIBZMQ */
	free(ctx.endline_str);
	argv_cleanup(args);

	if (res != nmsg_res_success || ctx.signal != 0) {
		if (ctx.debug >= 2) {
			if (ctx.signal == 0)
				fprintf(stderr, "%s: nmsg_io_loop() failed: %s (%d)\n", argv_program, nmsg_res_lookup(res),
					res);
			else
				fprintf(stderr, "%s: received signal: %s\n", argv_program, strsignal(ctx.signal));
		}
		return EXIT_FAILURE;
	}
	return EXIT_SUCCESS;
}

void
usage(const char *msg) {
	if (msg != NULL)
		fprintf(stderr, "%s: usage error: %s\n", argv_program, msg);
	else
		argv_usage(args, ARGV_USAGE_DEFAULT);

	for (size_t i = 0; i < statsmod_vec_size(ctx.statsmods_loaded); i++) {
		nmsg_statsmod_destroy(&statsmod_vec_data(ctx.statsmods_loaded)[i]);
	}
	statsmod_vec_destroy(&ctx.statsmods_loaded);

	nmsg_io_destroy(&ctx.io);
	exit(msg == NULL ? EXIT_SUCCESS : EXIT_FAILURE);
}

void
setup_nmsg_output(nmsgtool_ctx *c, nmsg_output_t output) {
	nmsg_output_set_buffered(output, !(c->unbuffered));
	nmsg_output_set_endline(output, c->endline_str);
	nmsg_output_set_zlibout(output, c->zlibout);
	nmsg_output_set_source(output, c->set_source);
	nmsg_output_set_operator(output, c->set_operator);
	nmsg_output_set_group(output, c->set_group);
}

void
setup_nmsg_input(nmsgtool_ctx *c, nmsg_input_t input) {
	if (c->vid != 0 && c->msgtype != 0)
		nmsg_input_set_filter_msgtype(input, c->vid, c->msgtype);
	nmsg_input_set_filter_source(input, c->get_source);
	nmsg_input_set_filter_operator(input, c->get_operator);
	nmsg_input_set_filter_group(input, c->get_group);
}

static void
print_io_stats(nmsg_io_t io) {
	uint64_t sum_in = 0, sum_out = 0, container_drops = 0, container_recvs = 0;

	if (nmsg_io_get_stats(io, &sum_in, &sum_out, &container_recvs, &container_drops) == nmsg_res_success)
		fprintf(stderr,
			"%s: totals: payloads_in %"PRIu64
			" payloads_out %"PRIu64
			" container_recvs %"PRIu64
		        " container_drops %"PRIu64"\n",
			argv_program, sum_in, sum_out, container_recvs, container_drops);
}

static void
io_close(struct nmsg_io_close_event *ce) {
	struct kickfile *kf;

	if (ctx.debug >= 2) {
		if (ce->close_type != nmsg_io_close_type_eof &&
		    ce->user != NULL && ce->user == ctx.stats_user) {
			print_io_stats(ce->io);
		}
	}

	if (ctx.debug >= 5) {
		fprintf(stderr, "entering io_close()\n");
		fprintf(stderr, "%s: ce->io_type = %u\n", __func__, ce->io_type);
		fprintf(stderr, "%s: ce->close_type = %u\n", __func__, ce->close_type);
		fprintf(stderr, "%s: ce->user = %p\n", __func__, ce->user);
		if (ce->io_type == nmsg_io_io_type_input) {
			fprintf(stderr, "%s: ce->input_type = %u\n", __func__, ce->input_type);
			fprintf(stderr, "%s: ce->input = %p\n", __func__, (void *)ce->input);
		} else if (ce->io_type == nmsg_io_io_type_output) {
			fprintf(stderr, "%s: ce->output_type = %u\n", __func__, ce->output_type);
			fprintf(stderr, "%s: ce->output = %p\n", __func__, (void *)ce->output);
		}
	}

	if (ce->user != NULL && ce->user != (void *) -1 &&
	    ce->io_type == nmsg_io_io_type_output &&
	    (ce->output_type == nmsg_output_type_stream ||
	     ce->output_type == nmsg_output_type_pres ||
	     ce->output_type == nmsg_output_type_json))
	{
		nmsg_output_close(ce->output);

		kf = (struct kickfile *) ce->user;
		kickfile_exec(kf);
		if (ce->close_type == nmsg_io_close_type_eof) {
			if (ctx.debug >= 2)
				fprintf(stderr, "%s: closed output: %s\n",
					argv_program, kf->basename);
			kickfile_destroy(&kf);
		} else {
			kickfile_rotate(kf);

			char *output_type_descr = NULL;
			switch (ce->output_type) {
			case nmsg_output_type_stream:
				*(ce->output) = nmsg_output_open_file(
					open_wfile(kf->tmpname), NMSG_WBUFSZ_MAX);
				output_type_descr = "nmsg";
				break;
			case nmsg_output_type_pres:
				*(ce->output) = nmsg_output_open_pres(
					open_wfile(kf->tmpname));
				output_type_descr = "pres";
				break;
			case nmsg_output_type_json:
				*(ce->output) = nmsg_output_open_json(
					open_wfile(kf->tmpname));
				output_type_descr = "json";
				break;
			default:
				assert(0);
			}
			setup_nmsg_output(&ctx, *(ce->output));
			if (ctx.debug >= 2)
				fprintf(stderr,
					"%s: reopened %s file output: %s\n",
					argv_program, output_type_descr, kf->curname);
		}
	} else if (ce->io_type == nmsg_io_io_type_input) {
		if ((ce->user == NULL || ce->close_type == nmsg_io_close_type_eof) &&
		     ce->input != NULL)
		{
			if (ctx.debug >= 5) {
				fprintf(stderr, "%s: closing input %p\n", __func__, (void *)ce->input);
			}
			nmsg_input_close(ce->input);
		}
	} else if (ce->io_type == nmsg_io_io_type_output) {
		if ((ce->user == NULL || ce->close_type == nmsg_io_close_type_eof) &&
		     ce->output != NULL)
		{
			if (ctx.debug >= 5) {
				fprintf(stderr, "%s: closing output %p\n", __func__, (void *)ce->output);
			}
			nmsg_output_close(ce->output);
		}
	} else {
		/* should never be reached */
		assert(0);
	}
}

static void
signal_handler(int sig) {
	fprintf(stderr, "%s: signalled break\n", argv_program);
	ctx.signal = sig;
	nmsg_io_breakloop(ctx.io);
}

static void
setup_signals(void) {
	struct sigaction sa;

	memset(&sa, 0, sizeof(sa));
	sigemptyset(&sa.sa_mask);
	sigaddset(&sa.sa_mask, SIGINT);
	sigaddset(&sa.sa_mask, SIGTERM);
	sa.sa_handler = &signal_handler;
	if (sigaction(SIGINT, &sa, NULL) != 0) {
		perror("sigaction");
		exit(EXIT_FAILURE);
	}
	if (sigaction(SIGTERM, &sa, NULL) != 0) {
		perror("sigaction");
		exit(EXIT_FAILURE);
	}

	ctx.signal = 0;
}
